File: 2.OOP/concepts/MediumExample/design_patterns/Structural/Adapter/Adapter.java
interface Shape{
    double getArea();
}
interface AreaMeasurable {
  double getAreaInCm();
}
class Square implements Shape{
    private final double side;

    public Square(double side){
        this.side = side;
    }
    @Override
    public double getArea(){
        return this.side * this.side;
    }
}
class Circle implements Shape{
    private final double radius;
    public Circle(double radius){
       this.radius = radius;
    }
    @Override
    public double getArea() {
      return Math.PI * radius * radius;
    }
}
// Adaptor Class : 
class ShapeAreaAdapter  implements AreaMeasurable{
    private final Shape  shape;
    public ShapeAreaAdapter (Shape  shape){
        this.shape = shape;
    }
    @Override
    public double getAreaInCm() {
        double areaInInches = this.shape.getArea();
        return areaInInches * 2.54; 
    }
}
// Usage Example 
public class AdapterDemo {
  static void displayAreaInCm(AreaMeasurable measurable) {
    System.out.println("Area in centimeters: " + measurable.getAreaInCm());
  }

  public static void main(String[] args) {
    Square square = new Square(5);
    Circle circle = new Circle(3);

    ShapeAreaAdapter squareAdapter = new ShapeAreaAdapter(square);
    ShapeAreaAdapter circleAdapter = new ShapeAreaAdapter(circle);

    displayAreaInCm(squareAdapter);
    displayAreaInCm(circleAdapter);
  }
}

File: 6.Software-Design-Patterns/example/Structural/Adapter/Adapter.go
package main

import (
	"fmt"
	"math"
)

type Shape interface {
	getArea() float64
}

type AreaMeasurable interface {
	getAreaInCm() float64
}

type Square struct {
	Side float64
}

func (this *Square) GetArea() float64 {
	return this.Side * this.Side
}

type Circle struct {
	Radius float64
}

func (this *Circle) GetArea() float64 {
	return math.Pi * this.Radius * this.Radius
}

type ShapeAreaCalculator struct {
	shape Shape
}

func (this *ShapeAreaCalculator) getAreaInCm() float64 {
	return this.shape.getArea() * 2.54
}
func main() {
	square := Square{Side: 5}
	circle := Circle{Radius: 3}
	fmt.Printf("Area of square: %.4f\n", square.GetArea())
	fmt.Printf("Area of circle: %.4f", circle.GetArea())
}
